core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      for (ViewVersion version : versionsToIndex) {
+        builder.put(version.versionId(), version);
+      }
+
+      return builder.build();
+    }
+
+    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
+      ImmutableMap.Builder<Integer, Schema> builder = ImmutableMap.builder();
+      for (Schema schema : schemasToIndex) {
+        builder.put(schema.schemaId(), schema);
+      }
+
+      return builder.build();
+    }
+
+    public ViewMetadata build() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
+      Preconditions.checkArgument(history.size() > 0, "Invalid view history: empty");
+      Preconditions.checkArgument(schemas.size() > 0, "Invalid schemas: empty");
+
+      Map<Integer, ViewVersion> versionsById = indexVersions(versions);
+      Preconditions.checkArgument(
+          versionsById.containsKey(currentVersionId),
+          "Cannot find current version %s in view versions: %s",
+          currentVersionId,
+          versionsById.keySet());
+
+      int currentSchemaId = versionsById.get(currentVersionId).schemaId();
+      Map<Integer, Schema> schemasById = indexSchemas(schemas);
+      Preconditions.checkArgument(
+          schemasById.containsKey(currentSchemaId),
+          "Cannot find current schema with id %s in schemas: %s",
+          currentSchemaId,
+          schemasById.keySet());
+
+      int versionHistorySizeToKeep =
+          PropertyUtil.propertyAsInt(
+              properties,
+              ViewProperties.VERSION_HISTORY_SIZE,
+              ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+
+      if (versionHistorySizeToKeep <= 0) {
+        LOG.warn(
+            "{} must be positive but was {}",
+            ViewProperties.VERSION_HISTORY_SIZE,
+            versionHistorySizeToKeep);
+      } else if (versions.size() > versionHistorySizeToKeep) {
+        List<ViewVersion> versionsToKeep =
+            versions.subList(versions.size() - versionHistorySizeToKeep, versions.size());
+        List<ViewHistoryEntry> historyToKeep =
+            history.subList(history.size() - versionHistorySizeToKeep, history.size());
+        List<MetadataUpdate> changesToKeep = Lists.newArrayList(changes);
+        Set<MetadataUpdate.AddViewVersion> toRemove =
+            changesToKeep.stream()
+                .filter(update -> update instanceof MetadataUpdate.AddViewVersion)
+                .map(update -> (MetadataUpdate.AddViewVersion) update)
+                .filter(
+                    update ->
+                        update.viewVersion().versionId() != currentVersionId
+                            && !versionsToKeep.contains(update.viewVersion()))
+                .collect(Collectors.toSet());
+        changesToKeep.removeAll(toRemove);
+
+        return ImmutableViewMetadata.of(
```

Comment:
We usually prefer having fewer return statements and using a variable like `outputVersions` or `finalHistory` that gets updated depending on the path.

Reply:
done